usr.sbin/daemon/daemon.c

```diff
@@ -304,246 +307,218 @@ main(int argc, char *argv[])
-				warn("sigprocmask");
-				daemon_terminate(state);
-			}
-			while (!terminate && !child_gone) {
-				sigsuspend(&state->mask_orig);
-			}
-			if (sigprocmask(SIG_UNBLOCK, &state->mask_susp, NULL)) {
-				warn("sigprocmask");
-				daemon_terminate(state);
-			}
+	while (state->mode != MODE_NOCHILD) {
+		printf("-- event loop tick.\n");
+		ret = kevent(kq, NULL, 0, &event, 1, NULL);
+		switch (ret) {
+		case -1:
+			err(EXIT_FAILURE, "kevent wait");
+		case 0:
 			continue;
 		}
 
-		if (sigprocmask(SIG_BLOCK, &state->mask_read, NULL)) {
-			warn("sigprocmask");
-			daemon_terminate(state);
+		if (event.flags & EV_ERROR) {
+			errx(EXIT_FAILURE, "Event error: %s", strerror(event.data));
 		}
 
-		state->child_eof = !listen_child(state->pipe_fd[0], state);
+		switch(event.filter) {
+		case EVFILT_SIGNAL:
+
+			printf("got SIGNAL\n");
+			switch (event.ident) {
+			case SIGCHLD:
+				printf("got SIGCHILD\n");
+				if (daemon_is_child_dead(state)) {
+					/* child is dead, read all until EOF */
+					state->pid = -1;
+					state->mode = MODE_NOCHILD;
+					while(listen_child(state->pipe_fd[0], state));
+				}
+				continue;
+			case SIGTERM:
+				printf("got SIGTERM\n");
+				if (state->mode != MODE_RUNNING) {
+					/* user is impatient */
+					/* TODO: warn about repeated SIGTERM? */
+					printf("ignoring SIGTERM\n");
+					continue;
+				}
 
-		if (sigprocmask(SIG_UNBLOCK, &state->mask_read, NULL)) {
-			warn("sigprocmask");
-			daemon_terminate(state);
-		}
+				state->mode = MODE_TERMINATING;
+				state->restart_enabled = false;
+				if (state->pid > 0) {
+					kill(state->pid, SIGTERM);
+				}
+				/*
+				 * TODO set kevent timer to exit
+				 * unconditionally after some time
+				 */
+				continue;
+			case SIGHUP:
+				printf("got SIGHUP\n");
+				if (state->log_reopen && state->output_fd >= 0) {
+					reopen_log(state);
+				}
+				continue;
+			}
+			break;
 
-	}
+		case EVFILT_READ:
+			printf("got data in the pipe\n");
+			/*
+			 * detecting EOF is no longer necessary
+			 * if child closes the pipe daemon will stop getting
+			 * EVFILT_READ events
+			 */
 
-	/*
-	 * At the end of the loop the the child is already gone.
-	 * Block SIGTERM to avoid racing until the child is spawned.
-	 */
-	if (sigprocmask(SIG_BLOCK, &state->mask_term, NULL)) {
-		warn("sigprocmask");
-		daemon_terminate(state);
+			if (event.data > 0) {
+				printf("got %ld bytes\n", event.data);
+				(void)listen_child(state->pipe_fd[0], state);
+			}
+			continue;
+		default:
+			continue;
+		}
 	}
 
-	/* sleep before exiting mainloop if restart is enabled */
-	if (state->restart_enabled && !terminate) {
-		daemon_sleep(state->restart_delay, 0);
-	}
+	printf("stopping event loop\n");
+	close(kq);
+	close(state->pipe_fd[0]);
+	state->pipe_fd[0] = -1;
 }
 
 static void
-daemon_sleep(time_t secs, long nsecs)
+daemon_sleep(struct daemon_state *state)
 {
-	struct timespec ts = { secs, nsecs };
+	struct timespec ts = { state->restart_delay, 0 };
```

Comment:
this is very clever.

will kqueue… queue events for us long enough?

i often set the restart delay to 30 seconds

Reply:
I dunno why `nanosleep` was used. It was always using 0 nanoseconds.
In the future this can probably be replaced with sleep()...

Reply:
Or better with timers+kq as you point out